crates/compiler/constrain/src/pattern.rs

```diff
@@ -183,6 +204,32 @@ fn headers_from_annotation_help(
     }
 }
 
+fn constrain_pattern_symbol(
+    types: &mut Types,
+    constraints: &mut Constraints,
+    region: Region,
+    expected: PExpectedTypeIndex,
+    state: &mut PatternState,
+    symbol: Symbol,
+) {
+    let expected = &constraints[expected];
+    let type_index = *expected.get_type_ref();
+
+    if could_be_a_tag_union(types, type_index) {
+        state
+            .delayed_is_open_constraints
+            .push(constraints.is_open_type(type_index));
+    }
```

Comment:
I don't think we want this for `as` patterns; this block is so that

```
when x is
  A -> ..
  b -> ..
```

is inferred as `[A]*`, and as patterns don't behave like that